package baidu.shuttle;
option cc_generic_services = true;

enum Status {
    kOk = 0;
    kGalaxyError = 1;
    kNoSuchJob = 2;
    kNoMore = 3;
    kOpenFileFail = 5;
    kCloseFileFail = 6;
    kWriteFileFail = 7;
    kInvalidArg = 8;
    kBadMagic = 9;
    kReadFileFail = 10;
    kNotImplement = 11;
    kNoSuchTask = 12;
    kSuspend = 13;
    kUnKnown = 20;
}

enum JobPriority {
    kVeryHigh = 0;
    kHigh = 1;
    kNormal = 2;
    kLow = 3;
}

enum TaskState {
    kTaskPending = 0;
    kTaskRunning = 1;
    kTaskFailed = 2;
    kTaskKilled = 3;
    kTaskCompleted = 4;
    kTaskCanceled = 5;
    kTaskMoveOutputFailed = 6;
    kTaskUnknown = 10;
}

enum Partition {
    kKeyFieldBasedPartitioner = 0;
    kIntHashPartitioner = 1;
}

enum WorkMode {
    kMap = 0;
    kReduce = 1;
    kMapOnly = 2;
}

enum JobType {
    kMapOnlyJob = 0;
    kMapReduceJob = 1;
}

message DfsInfo {
    optional string host = 1;
    optional string port = 2;
    optional string user = 3;
    optional string password = 4;    
}

enum InputFormat {
    kTextInput = 0;
    kBinaryInput = 1;
    kNLineInput = 2;
}

enum OutputFormat {
    kTextOutput = 0;
    kBinaryOutput = 1;
    kSuffixMultipleTextOutput = 2;
}

enum PipeStyle {
    kStreaming = 0;
    kBiStreaming = 1;
}

message NodeConfig {
    optional int32 node = 1;
    optional int32 capacity = 2;
    optional int32 total = 3;
    optional int32 millicores = 4;
    optional int64 memory = 5;
    optional string command = 6;
    repeated string inputs = 7;
    optional InputFormat input_format = 8;
    optional DfsInfo input_dfs = 9;
    optional string output = 10;
    optional OutputFormat output_format = 11;
    optional DfsInfo output_dfs = 12;
    optional Partition partition = 13;
    optional string key_separator = 14;
    optional int32 key_fields_num = 15;
    optional int32 partition_fields_num = 16;
    optional bool allow_duplicates = 17 [default = true];
    optional int32 retry = 18;
}

message JobDescriptor {
    optional string name = 1;
    optional string user = 2;
    optional JobPriority priority = 3;
    repeated string files = 4;
    optional PipeStyle pipe_style = 5;
    optional int64 split_size = 6;
    repeated NodeConfig nodes = 7;
    message NodeNeigbor {
        repeated int32 neigbors = 1;
    }
    repeated NodeNeigbor map = 8;
}

message TaskInput {
    optional string input_file = 3;
    optional int64 input_offset = 4;
    optional int64 input_size = 5;
}

message TaskInfo {
    optional int32 task_id = 1;
    optional int32 attempt_id = 2;
    optional TaskInput input = 3;
    optional WorkMode task_type = 4;
    optional JobDescriptor job = 5;
}
